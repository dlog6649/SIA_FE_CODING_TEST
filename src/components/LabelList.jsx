import React, { useEffect, useRef } from 'react';
import { LABEL_CREATE_MODE } from '../modules/annotator';


const compareIds = (_ids) => {
    let ids = [];
    document.querySelectorAll('.label-info.active').forEach(labelInfo => {
        ids.push(parseInt(labelInfo.dataset.id));
    });
    if (ids.length !== _ids.length) {
        return false;
    }
    for (let i = 0; i < _ids.length; i++) {
        if (parseInt(_ids[i]) !== parseInt(ids[i])) {
            return false;
        }
    }
    return true;
}


export default function LabelList(props) {

    const refLabelList = useRef(null);

    useEffect(() => {
        console.log('LabelList useEffect [props.labels]');
        let labels = props.labels;
        if (!labels) {
            labels = [];
        }

        let labelListRoot = refLabelList.current.lastChild;

        while(labelListRoot.firstChild) {
            labelListRoot.removeChild(labelListRoot.firstChild);
        }

        let labelList = '';

        labels.forEach(label => {
            labelList += `
                <li class="label-info btn" data-id="${label.id}" data-testid="testLabelInfo">
                    <p class="label-class">${label.name}</p>
                    <p class="label-coordinate">
                        (${parseInt(label.coordinates[0].x)}, ${parseInt(label.coordinates[0].y)})
                        (${parseInt(label.coordinates[1].x)}, ${parseInt(label.coordinates[1].y)})
                        (${parseInt(label.coordinates[2].x)}, ${parseInt(label.coordinates[2].y)})
                        (${parseInt(label.coordinates[3].x)}, ${parseInt(label.coordinates[3].y)})
                    </p>
                </li>
            `;
        });

        labelListRoot.insertAdjacentHTML('afterbegin', labelList);
        
    }, [props.labels]);


    useEffect(() => {
        console.log('LabelList useEffect [props.selectedLabelsIds]');

        if(compareIds(props.selectedLabelsIds)) {
            console.log('LabelList useEffect [props.selectedLabelsIds] returned');
            return;
        }

        document.querySelectorAll('.label-info').forEach(labelInfo => {
            labelInfo.classList.remove('active');

            props.selectedLabelsIds.forEach(id => {
                if(parseInt(labelInfo.dataset.id) === parseInt(id)) {
                    labelInfo.classList.add('active');
                }
            });
        });

    }, [props.selectedLabelsIds]);


    const toggleLabelList = () => {
        let controller = document.querySelector('.label-list-controller');
        let img = document.querySelector('.label-list-btn-img');
        let labelListRoot = document.querySelector('.label-list-root');

        if (labelListRoot.style.display === 'block') {
            controller.firstChild.style.display = 'none';
            controller.style.minWidth = '38px';
            controller.style.borderRight = '1px solid lightgray';
            img.src = require('../asset/images/arrow-right.png');
            labelListRoot.style.display = 'none';
            labelListRoot.parentNode.style.borderRight = 'none';
        }
        else {
            controller.firstChild.style.display = 'block';
            controller.style.minWidth = '300px';
            controller.style.borderRight = 'none';
            img.src = require('../asset/images/arrow-left.png');
            labelListRoot.style.display = 'block';
            labelListRoot.parentNode.style.borderRight = '1px solid lightgray';
        }
    };
    

    const selectLabel = e => {
        if(props.mode === LABEL_CREATE_MODE) {
            return;
        }
        
        let _labelInfo;
        _labelInfo = e.target.classList.contains('label-info') ? e.target : e.target.parentNode;
        _labelInfo.classList.add('active');

        if(!e.ctrlKey) {
            document.querySelectorAll('.label-info.active').forEach(labelInfo => {
                if(_labelInfo === labelInfo) {
                    return true;
                }
                labelInfo.classList.remove('active');
            });
        }

        let ids = [];
        document.querySelectorAll('.label-info.active').forEach(labelInfo => {
            ids.push(parseInt(labelInfo.dataset.id));
        });

        props.selectLabels(ids);
    }


    return (
        <div className="label-list" ref={refLabelList}>
            <div className="label-list-controller">
                <span>Labels</span>
                <button className="btn label-list-btn" onClick={toggleLabelList} type="button">
                    <img className="label-list-btn-img" src={require('../asset/images/arrow-left.png')} alt="arrow-left"/>
                </button>
            </div>
            <ul className="label-list-root" style={{display:'block'}} onMouseDown={selectLabel}>
            </ul>
        </div>
    );
}


