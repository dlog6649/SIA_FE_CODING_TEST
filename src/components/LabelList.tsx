import React, { useEffect, useRef } from "react";
import { useDispatch } from "react-redux";
import { LABEL_CREATE_MODE } from "../modules/annotator";
import imgArrowRight from "../asset/images/arrow-right.png";
import imgArrowLeft from "../asset/images/arrow-left.png";

const compareIds = (_ids: Array<number>) => {
  const ids = [] as Array<number>;
  document.querySelectorAll(".label-info.active").forEach((labelInfo: Element) => {
    const info: HTMLLIElement = labelInfo as HTMLLIElement;
    ids.push(Number(info.dataset.id));
  });
  if (ids.length !== _ids.length) {
    return false;
  }
  for (let i = 0; i < _ids.length; i++) {
    if (Number(_ids[i]) !== Number(ids[i])) {
      return false;
    }
  }
  return true;
};

interface Label {
  id: number;
  name: string;
  coordinates: Array<{ x: number; y: number }>;
  data: { x: number; y: number; w: number; h: number; deg: number };
}

interface Props {
  mode: string;
  labels: Array<Label>;
  selectedLabelsIds: Array<number>;
  selectLabels: (selectedLabelsIds: Array<number>) => void;
}

export default function LabelList(props: Props) {
  const refLabelList = useRef<HTMLDivElement>(null);

  useEffect(() => {
    console.log("LabelList useEffect [props.labels]");
    let { labels } = props;
    if (!labels) {
      labels = [] as Array<Label>;
    }
    if (!refLabelList.current) {
      return;
    }
    const curRefLabelList = refLabelList.current;
    const labelListRoot = curRefLabelList.lastChild as HTMLUListElement;
    while (labelListRoot.firstChild) {
      labelListRoot.removeChild(labelListRoot.firstChild);
    }
    let labelList = "";
    labels.forEach((label) => {
      labelList += `
                <li class="label-info btn" data-id="${label.id}" data-testid="testLabelInfo">
                    <p class="label-class">${label.name}</p>
                    <p class="label-coordinate">
                        (${Number(label.coordinates[0].x)}, ${Number(label.coordinates[0].y)})
                        (${Number(label.coordinates[1].x)}, ${Number(label.coordinates[1].y)})
                        (${Number(label.coordinates[2].x)}, ${Number(label.coordinates[2].y)})
                        (${Number(label.coordinates[3].x)}, ${Number(label.coordinates[3].y)})
                    </p>
                </li>
            `;
    });

    labelListRoot.insertAdjacentHTML("afterbegin", labelList);
  }, [props.labels]);

  useEffect(() => {
    console.log("LabelList useEffect [props.selectedLabelsIds]");
    if (compareIds(props.selectedLabelsIds)) {
      console.log("LabelList useEffect [props.selectedLabelsIds] returned");
      return;
    }
    document.querySelectorAll(".label-info").forEach((labelInfo) => {
      const info: HTMLElement = labelInfo as HTMLElement;
      info.classList.remove("active");
      props.selectedLabelsIds.forEach((id) => {
        if (Number(info.dataset.id) === Number(id)) {
          info.classList.add("active");
        }
      });
    });
  }, [props.selectedLabelsIds]);

  const toggleLabelList = () => {
    const controller = document.querySelector(".label-list-controller") as HTMLDivElement;
    const img = document.querySelector(".label-list-btn-img") as HTMLImageElement;
    const labelListRoot = document.querySelector(".label-list-root") as HTMLUListElement;
    if (labelListRoot.style.display === "block") {
      (controller.firstChild as HTMLElement).style.display = "none";
      controller.style.minWidth = "38px";
      controller.style.borderRight = "1px solid lightgray";
      img.src = imgArrowRight;
      labelListRoot.style.display = "none";
      (labelListRoot.parentNode as HTMLElement).style.borderRight = "none";
    } else {
      (controller.firstChild as HTMLElement).style.display = "block";
      controller.style.minWidth = "300px";
      controller.style.borderRight = "none";
      img.src = imgArrowLeft;
      labelListRoot.style.display = "block";
      (labelListRoot.parentNode as HTMLElement).style.borderRight = "1px solid lightgray";
    }
  };

  const selectLabel = (evt: any): void => {
    if (props.mode === LABEL_CREATE_MODE) {
      return;
    }
    const labelInfo: HTMLLIElement = evt.target.classList.contains("label-info") ? evt.target : evt.target.parentNode;
    labelInfo.classList.add("active");
    if (!evt.ctrlKey) {
      document.querySelectorAll(".label-info.active").forEach((info: Element) => {
        if (labelInfo !== info) {
          info.classList.remove("active");
        }
      });
    }
    const ids = [] as Array<number>;
    document.querySelectorAll(".label-info.active").forEach((info: Element) => {
      ids.push(Number((info as HTMLLIElement).dataset.id as string));
    });
    props.selectLabels(ids);
  };

  return (
    <div className="label-list" ref={refLabelList}>
      <div className="label-list-controller">
        <span>Labels</span>
        <button className="btn label-list-btn" onClick={toggleLabelList} type="button">
          <img className="label-list-btn-img" src={imgArrowLeft} alt="arrow-left" />
        </button>
      </div>
      <ul className="label-list-root" style={{ display: "block" }} onMouseDown={selectLabel} />
    </div>
  );
}
