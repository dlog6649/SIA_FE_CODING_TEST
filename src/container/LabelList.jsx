import LabelList from '../components/LabelList';
import { connect } from 'react-redux';
import { selectLabels } from '../modules/annotator';

const mapStateToProps = state => {
    let _labels = state.annotator.labels.filter(label => label.url === state.annotator.curImgURL);
    return {
        labels: _labels,
        selectedLabelIds: state.annotator.selectedLabelIds
    };
}
const mapDispatchToProps = dispatch => {
    return {
        selectLabels: ids => dispatch(selectLabels(ids))
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(LabelList);