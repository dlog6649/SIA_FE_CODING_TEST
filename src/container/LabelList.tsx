import { connect } from "react-redux";
import LabelList from "../components/LabelList";
import { selectLabels, AnnotatorState } from "../modules/annotator";

interface State {
  annotator: AnnotatorState;
}

const mapStateToProps = (state: State) => {
  return {
    mode: state.annotator.mode,
    labels: state.annotator.labels[state.annotator.currentImgURL],
    selectedLabelsIds: state.annotator.selectedLabelsIds,
  };
};

const mapDispatchToProps = (dispatch: any) => {
  return {
    selectLabels: (selectedLabelsIds: Array<number>) => dispatch(selectLabels(selectedLabelsIds)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(LabelList);
