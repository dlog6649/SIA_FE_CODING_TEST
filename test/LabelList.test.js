import React from "react";
import { render, fireEvent } from "@testing-library/react";
import LabelList from "./LabelList";

describe("LabelList 테스트 시작", () => {
  let wrapper = null;
  const _labels = [
    {
      id: 0,
      name: "sea",
      coordinates: [
        { x: 10, y: 20 },
        { x: 110, y: 20 },
        { x: 110, y: 70 },
        { x: 10, y: 70 },
      ],
      data: { x: 10, y: 20, w: 90, h: 50, deg: 0 },
    },
    {
      id: 1,
      name: "woods",
      coordinates: [
        { x: 10, y: 20 },
        { x: 110, y: 20 },
        { x: 110, y: 70 },
        { x: 10, y: 70 },
      ],
      data: { x: 10, y: 20, w: 90, h: 50, deg: 0 },
    },
    {
      id: 2,
      name: "building",
      coordinates: [
        { x: 10, y: 20 },
        { x: 110, y: 20 },
        { x: 110, y: 70 },
        { x: 10, y: 70 },
      ],
      data: { x: 10, y: 20, w: 90, h: 50, deg: 0 },
    },
  ];
  const _selectedLabelsIds = [1];
  const mockSelectLabels = jest.fn();

  it("렌더링된 스냅샷이 기존과 일치해야 한다.", () => {
    wrapper = render(<LabelList labels={_labels} selectedLabelsIds={_selectedLabelsIds} selectLabels={mockSelectLabels} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("선택된 라벨ID가 전달되었을 때 목록에 활성화가 되어야 한다.", () => {
    // 선택된 라벨ID로 1번 전달
    wrapper = render(<LabelList labels={_labels} selectedLabelsIds={[1]} selectLabels={mockSelectLabels} />);

    const labelInfo = wrapper.getAllByTestId("testLabelInfo");
    expect(labelInfo[1].dataset.id).toEqual("1");
    expect(labelInfo[1].classList.contains("active")).toBe(true);
  });

  it("목록에 있는 라벨정보를 선택시 해당 라벨들에 대한 ID리스트가 디스패치돼야 한다.", () => {
    wrapper = render(<LabelList labels={_labels} selectedLabelsIds={[]} selectLabels={mockSelectLabels} />);
    const labelInfo = wrapper.getAllByTestId("testLabelInfo");

    // 라벨목록중 2번 클릭
    fireEvent.mouseDown(labelInfo[2]);

    const selectLabelsEventByClick = mockSelectLabels.mock.calls;
    const [selectedLabelsIds] = selectLabelsEventByClick[0];
    expect(selectedLabelsIds).toEqual([2]);
  });
});
