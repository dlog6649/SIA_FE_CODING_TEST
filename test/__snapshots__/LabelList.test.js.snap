// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelList 테스트 시작 렌더링된 스냅샷이 기존과 일치해야 한다. 1`] = `
<body>
  <div>
    <div
      class="label-list"
    >
      <div
        class="label-list-controller"
      >
        <span>
          Labels
        </span>
        <button
          class="btn label-list-btn"
          type="button"
        >
          <img
            alt="arrow-left"
            class="label-list-btn-img"
            src="arrow-left.png"
          />
        </button>
      </div>
      <ul
        class="label-list-root"
        style="display: block;"
      >
        
                
        <li
          class="label-info btn"
          data-id="0"
          data-testid="testLabelInfo"
        >
          
                    
          <p
            class="label-class"
          >
            sea
          </p>
          
                    
          <p
            class="label-coordinate"
          >
            
                        (10, 20)
                        (110, 20)
                        (110, 70)
                        (10, 70)
                    
          </p>
          
                
        </li>
        
            
                
        <li
          class="label-info btn active"
          data-id="1"
          data-testid="testLabelInfo"
        >
          
                    
          <p
            class="label-class"
          >
            woods
          </p>
          
                    
          <p
            class="label-coordinate"
          >
            
                        (10, 20)
                        (110, 20)
                        (110, 70)
                        (10, 70)
                    
          </p>
          
                
        </li>
        
            
                
        <li
          class="label-info btn"
          data-id="2"
          data-testid="testLabelInfo"
        >
          
                    
          <p
            class="label-class"
          >
            building
          </p>
          
                    
          <p
            class="label-coordinate"
          >
            
                        (10, 20)
                        (110, 20)
                        (110, 70)
                        (10, 70)
                    
          </p>
          
                
        </li>
        
            
      </ul>
    </div>
  </div>
</body>
`;
